pkg/volume/host_path/host_path.go

```diff
@@ -308,3 +323,152 @@ func getVolumeSource(spec *volume.Spec) (*v1.HostPathVolumeSource, bool, error)
 
 	return nil, false, fmt.Errorf("Spec does not reference an HostPath volume type")
 }
+
+type hostPathTypeChecker interface {
+	Exists() bool
+	IsFile() bool
+	MakeFile() error
+	IsDir() bool
+	MakeDir() error
+	IsBlock() bool
+	IsChar() bool
+	IsSocket() bool
+}
+
+func getFileType(fileInfo os.FileInfo) (v1.HostPathType, error) {
+	mode := fileInfo.Sys().(*syscall.Stat_t).Mode
+	switch mode & syscall.S_IFMT {
+	case syscall.S_IFSOCK:
+		return v1.HostPathSocket, nil
+	case syscall.S_IFBLK:
+		return v1.HostPathBlockDev, nil
+	case syscall.S_IFCHR:
+		return v1.HostPathCharDev, nil
+	}
+
+	return "", fmt.Errorf("only recongnize socket, block device and character device")
+}
+
+type osFileTypeChecker struct {
+	path   string
+	exists bool
+	info   os.FileInfo
+}
+
+func (ftc osFileTypeChecker) Exists() bool {
+	return ftc.exists
+}
+
+func (ftc osFileTypeChecker) IsFile() bool {
+	if !ftc.Exists() {
+		return false
+	}
+
+	return !ftc.info.IsDir()
+}
+
+func (ftc osFileTypeChecker) MakeFile() error {
+	f, err := os.Create(ftc.path)
+	defer f.Close()
+	return err
+}
+
+func (ftc osFileTypeChecker) IsDir() bool {
+	if !ftc.Exists() {
+		return false
+	}
+	return ftc.info.IsDir()
+}
+
+func (ftc osFileTypeChecker) MakeDir() error {
+	err := os.Mkdir(ftc.path, os.FileMode(0644))
+	return err
+}
+
+func (ftc osFileTypeChecker) IsBlock() bool {
+	if !ftc.Exists() {
+		return false
+	}
+
+	blkDevType, err := getFileType(ftc.info)
+	if err != nil {
+		return false
+	}
+	return blkDevType == v1.HostPathBlockDev
+}
+
+func (ftc osFileTypeChecker) IsChar() bool {
+	if !ftc.Exists() {
+		return false
+	}
+
+	charDevType, err := getFileType(ftc.info)
+	if err != nil {
+		return false
+	}
+	return charDevType == v1.HostPathCharDev
+}
+
+func (ftc osFileTypeChecker) IsSocket() bool {
+	if !ftc.Exists() {
+		return false
+	}
+
+	socketType, err := getFileType(ftc.info)
+	if err != nil {
+		return false
+	}
+	return socketType == v1.HostPathSocket
+}
+
+func newOSFileTypeChecker(path string) (hostPathTypeChecker, error) {
+	ftc := osFileTypeChecker{path: path}
+	info, err := os.Stat(path)
+	if err != nil {
+		ftc.info = info
+	}
+	ftc.exists = os.IsExist(err)
+	return ftc, err
+}
+
+func checkType(path string, pathType *v1.HostPathType) error {
+	ftc, _ := newOSFileTypeChecker(path)
+	switch *pathType {
+	case v1.HostPathDirectoryOrCreate:
+		if !ftc.Exists() {
```

Comment:
This doesn't actually check for directory-ness.  Did you mean to fall-through?  You have to say that in go.